flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateReduceGroupFunction.scala

```diff
@@ -54,44 +55,51 @@ class AggregateReduceGroupFunction(
   override def open(config: Configuration) {
     Preconditions.checkNotNull(aggregates)
     Preconditions.checkNotNull(groupKeysMapping)
-    aggregateBuffer = new Row(intermediateRowArity)
+    aggregateBuffer = new Row(aggregates.length + groupKeysMapping.length)
     output = new Row(finalRowArity)
     if (!groupingSetsMapping.isEmpty) {
       intermediateGroupKeys = Some(groupKeysMapping.map(_._1))
     }
   }
 
   /**
-   * For grouped intermediate aggregate Rows, merge all of them into aggregate buffer,
-   * calculate aggregated values output by aggregate buffer, and set them into output 
-   * Row based on the mapping relation between intermediate aggregate data and output data.
-   *
-   * @param records  Grouped intermediate aggregate Rows iterator.
-   * @param out The collector to hand results to.
-   *
-   */
+    * For grouped intermediate aggregate Rows, merge all of them into aggregate buffer,
+    * calculate aggregated values output by aggregate buffer, and set them into output
+    * Row based on the mapping relation between intermediate aggregate data and output data.
+    *
+    * @param records Grouped intermediate aggregate Rows iterator.
+    * @param out     The collector to hand results to.
+    *
+    */
   override def reduce(records: Iterable[Row], out: Collector[Row]): Unit = {
 
-    // Initiate intermediate aggregate value.
-    aggregates.foreach(_.initiate(aggregateBuffer))
-
-    // Merge intermediate aggregate value to buffer.
+    // merge intermediate aggregate value to buffer.
     var last: Row = null
-    records.foreach((record) => {
-      aggregates.foreach(_.merge(record, aggregateBuffer))
+    val iterator = records.iterator()
+    val accumulatorList = Array.fill(aggregates.length) {
+      new JArrayList[Accumulator]()
+    }
+
+    while (iterator.hasNext) {
+      val record = iterator.next()
+      for (i <- aggregates.indices) {
+        accumulatorList(i).add(
```

Comment:
I'm wondering whether it would be better to have no preparing mapper and instead have two separate path for combineable and non-combineable.
- Combinable: `input -> groupCombine() -> groupReduce()` where `groupCombine` uses `accumulate()` and `groupReduce` pairwise merges.
- Non-Combinable: `input -> groupReduce()` where `groupReduce` uses `accumulate()`

The combinable `groupCombine` and non-combinable `groupReduce` might even use the same code (except that `groupReduce` needs to call `getValue()` in the end).